File: examples/showcase/github_fronting/custom-code-bundles/github_fronting_typescript/src/runtime/backend-adapter.ts
import type { BackendInvocationPlan, GeneratedCapabilityRuntimeMetadata } from "../generated/runtime-target.js";

export type GeneratedBackendInvocationContext = {
  rootPrincipal?: string;
  approvalGrant?: string | null;
};

export interface GeneratedBackendAdapter {
  execute(
    capability: GeneratedCapabilityRuntimeMetadata,
    plan: BackendInvocationPlan,
    adapterInput: Record<string, unknown>,
    context: GeneratedBackendInvocationContext,
  ): Promise<Record<string, unknown>>;
}

const githubToken = () => process.env.GITHUB_TOKEN?.trim() || null;

function csvEnv(name: string): Set<string> {
  return new Set((process.env[name] ?? "").split(",").map((item) => item.trim().toLowerCase()).filter(Boolean));
}

function repoAllowed(owner: string, repo: string): boolean {
  const key = `${owner}/${repo}`.toLowerCase();
  const blocked = csvEnv("ANIP_GITHUB_BLOCKED_REPOS");
  const allowed = csvEnv("ANIP_GITHUB_ALLOWED_REPOS");
  if (blocked.has(key)) return false;
  return allowed.size === 0 || allowed.has(key);
}

function limitValue(value: unknown, fallback = 20, max = 50): number {
  const parsed = Number.parseInt(String(value ?? ""), 10);
  if (!Number.isFinite(parsed)) return fallback;
  return Math.max(1, Math.min(parsed, max));
}

function stringList(value: unknown): string[] {
  const raw = Array.isArray(value) ? value : typeof value === "string" && value.trim() ? value.split(",") : [];
  return Array.from(new Set(raw.map((item) => String(item).trim()).filter(Boolean)));
}

async function githubRequest(method: string, path: string, token: string, body?: Record<string, unknown>): Promise<Record<string, unknown>> {
  const response = await fetch(`https://api.github.com${path}`, {
    method,
    headers: {
      Accept: "application/vnd.github+json",
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
      "User-Agent": "anip-github-fronting-showcase",
      "X-GitHub-Api-Version": "2022-11-28",
    },
    body: body ? JSON.stringify(body) : undefined,
  });
  const text = await response.text();
  const payload = text ? JSON.parse(text) as Record<string, unknown> : {};
  if (!response.ok) return { error: "github_http_error", status: response.status, detail: payload };
  return payload;
}

function metadata(capability: GeneratedCapabilityRuntimeMetadata, plan: BackendInvocationPlan): Record<string, unknown> {
  return {
    capability_id: capability.capability_id,
    selected_backend: plan.selected_binding,
    semantic_input: plan.semantic_input,
  };
}

function restricted(capability: GeneratedCapabilityRuntimeMetadata, plan: BackendInvocationPlan, owner: string, repo: string): Record<string, unknown> {
  return {
    execution_status: "restricted",
    ...metadata(capability, plan),
    repository: { owner, repo },
    reason: "GitHub repository is outside the configured ANIP repository policy.",
  };
}

function mutationEnabled(context: GeneratedBackendInvocationContext): boolean {
  return process.env.ANIP_GITHUB_ALLOW_MUTATION === "true" && Boolean(context.approvalGrant);
}

function repoSummary(repoPayload: Record<string, unknown>): Record<string, unknown> {
  const owner = repoPayload.owner as Record<string, unknown> | undefined;
  return {
    owner: owner?.login,
    repo: repoPayload.name,
    default_branch: repoPayload.default_branch,
    private: repoPayload.private,
    html_url: repoPayload.html_url,
  };
}

function preview(
  capability: GeneratedCapabilityRuntimeMetadata,
  plan: BackendInvocationPlan,
  action: string,
  path: string,
  body: Record<string, unknown>,
  repoPayload: Record<string, unknown>,
): Record<string, unknown> {
  return {
    execution_status: "prepared",
    ...metadata(capability, plan),
    approval_required: true,
    mutation_performed: false,
    github_action: action,
    github_metadata: repoSummary(repoPayload),
    github_request: { method: "POST", path, body },
    note: "Prepared a GitHub request payload. No GitHub mutation was performed.",
  };
}

async function repoMetadata(
  capability: GeneratedCapabilityRuntimeMetadata,
  plan: BackendInvocationPlan,
  params: Record<string, unknown>,
  token: string,
): Promise<{ owner: string; repo: string; repoPayload?: Record<string, unknown>; error?: Record<string, unknown> }> {
  const owner = String(params.owner ?? "").trim();
  const repo = String(params.repo ?? "").trim();
  if (!repoAllowed(owner, repo)) return { owner, repo, error: restricted(capability, plan, owner, repo) };
  const repoPayload = await githubRequest("GET", `/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}`, token);
  if (repoPayload.error) return { owner, repo, error: { execution_status: "backend_error", ...metadata(capability, plan), github_error: repoPayload } };
  return { owner, repo, repoPayload };
}

export function createDefaultBackendAdapter(): GeneratedBackendAdapter {
  return {
    async execute(capability, plan, adapterInput, context) {
      if (plan.unresolved_required_backend_inputs.length > 0) {
        return {
          execution_status: "backend_input_incomplete",
          capability_id: capability.capability_id,
          backend_input_contract: plan.backend_input_contract,
          unresolved_required_backend_inputs: plan.unresolved_required_backend_inputs,
          note: "Generated host is runnable, but backend-only inputs still require extension completion.",
        };
      }
      const token = githubToken();
      if (!token) return { execution_status: "backend_error", ...metadata(capability, plan), github_error: { error: "missing_github_token" } };
      if (capability.capability_id === "github.repo.search_context") return searchRepositoryContext(capability, plan, adapterInput, token);
      if (capability.capability_id === "github.issue.prepare") return prepareOrCreateIssue(capability, plan, adapterInput, token, context);
      if (capability.capability_id === "github.pr.comment.prepare") return preparePullRequestComment(capability, plan, adapterInput, token, context);
      if (capability.capability_id === "github.workflow.dispatch.request") return prepareWorkflowDispatch(capability, plan, adapterInput, token, context);
      if (capability.capability_id === "github.release_notes.prepare") return prepareReleaseNotes(capability, plan, adapterInput, token);
      return { execution_status: "backend_execution_stub", ...metadata(capability, plan) };
    },
  };
}

async function searchRepositoryContext(
  capability: GeneratedCapabilityRuntimeMetadata,
  plan: BackendInvocationPlan,
  params: Record<string, unknown>,
  token: string,
): Promise<Record<string, unknown>> {
  const owner = String(params.owner ?? "").trim();
  const repo = String(params.repo ?? "").trim();
  if (!repoAllowed(owner, repo)) return restricted(capability, plan, owner, repo);
  const query = String(params.query ?? "").trim();
  const limit = limitValue(params.limit);
  const githubQuery = `repo:${owner}/${repo} ${query}`.trim();
  const payload = await githubRequest("GET", `/search/issues?${new URLSearchParams({ q: githubQuery, per_page: String(limit) })}`, token);
  if (payload.error) return { execution_status: "backend_error", ...metadata(capability, plan), github_error: payload };
  const items = ((payload.items ?? []) as Record<string, unknown>[]).slice(0, limit).map((item) => ({
    number: item.number,
    title: item.title,
    state: item.state,
    html_url: item.html_url,
    kind: item.pull_request ? "pull_request" : "issue",
  }));
  return {
    execution_status: "completed",
    ...metadata(capability, plan),
    github_query: githubQuery,
    result: { items, count: items.length, total_count: payload.total_count },
  };
}

async function prepareOrCreateIssue(
  capability: GeneratedCapabilityRuntimeMetadata,
  plan: BackendInvocationPlan,
  params: Record<string, unknown>,
  token: string,
  context: GeneratedBackendInvocationContext,
): Promise<Record<string, unknown>> {
  const repo = await repoMetadata(capability, plan, params, token);
  if (repo.error) return repo.error;
  const body: Record<string, unknown> = { title: String(params.title ?? "").trim(), body: String(params.body ?? "").trim() };
  const labels = stringList(params.labels);
  const assignees = stringList(params.assignees);
  if (labels.length) body.labels = labels;
  if (assignees.length) body.assignees = assignees;
  const prepared = preview(capability, plan, "issues.create", `/repos/${repo.owner}/${repo.repo}/issues`, body, repo.repoPayload!);
  if (!mutationEnabled(context)) return prepared;
  const created = await githubRequest("POST", `/repos/${encodeURIComponent(repo.owner)}/${encodeURIComponent(repo.repo)}/issues`, token, body);
  if (created.error) return { ...prepared, execution_status: "backend_error", github_error: created };
  return {
    ...prepared,
    execution_status: "completed",
    approval_required: false,
    mutation_performed: true,
    created_issue: { number: created.number, html_url: created.html_url, state: created.state },
    note: "Created GitHub issue after the ANIP runtime validated and reserved an approval grant.",
  };
}

async function preparePullRequestComment(
  capability: GeneratedCapabilityRuntimeMetadata,
  plan: BackendInvocationPlan,
  params: Record<string, unknown>,
  token: string,
  context: GeneratedBackendInvocationContext,
): Promise<Record<string, unknown>> {
  const repo = await repoMetadata(capability, plan, params, token);
  if (repo.error) return repo.error;
  const pullNumber = String(params.pull_number ?? "").trim();
  const pull = await githubRequest("GET", `/repos/${encodeURIComponent(repo.owner)}/${encodeURIComponent(repo.repo)}/pulls/${encodeURIComponent(pullNumber)}`, token);
  if (pull.error) return { execution_status: "backend_error", ...metadata(capability, plan), github_error: pull };
  const body = { body: `[${String(params.comment_purpose ?? "triage_update").trim()}] ${String(params.context ?? "").trim()}`.trim() };
  const prepared = preview(capability, plan, "issues.createComment", `/repos/${repo.owner}/${repo.repo}/issues/${pullNumber}/comments`, body, repo.repoPayload!);
  prepared.pull_request = { number: pull.number, title: pull.title, state: pull.state };
  if (!mutationEnabled(context)) return prepared;
  const posted = await githubRequest("POST", `/repos/${encodeURIComponent(repo.owner)}/${encodeURIComponent(repo.repo)}/issues/${encodeURIComponent(pullNumber)}/comments`, token, body);
  if (posted.error) return { ...prepared, execution_status: "backend_error", github_error: posted };
  return { ...prepared, execution_status: "completed", approval_required: false, mutation_performed: true, posted_comment: { id: posted.id, html_url: posted.html_url } };
}

async function prepareWorkflowDispatch(
  capability: GeneratedCapabilityRuntimeMetadata,
  plan: BackendInvocationPlan,
  params: Record<string, unknown>,
  token: string,
  context: GeneratedBackendInvocationContext,
): Promise<Record<string, unknown>> {
  const repo = await repoMetadata(capability, plan, params, token);
  if (repo.error) return repo.error;
  const workflowId = String(params.workflow_id ?? "").trim();
  const body = { ref: String(params.ref ?? "").trim(), inputs: params.inputs ?? {} };
  const prepared = preview(capability, plan, "actions.createWorkflowDispatch", `/repos/${repo.owner}/${repo.repo}/actions/workflows/${workflowId}/dispatches`, body, repo.repoPayload!);
  if (!mutationEnabled(context)) return prepared;
  const dispatched = await githubRequest("POST", `/repos/${encodeURIComponent(repo.owner)}/${encodeURIComponent(repo.repo)}/actions/workflows/${encodeURIComponent(workflowId)}/dispatches`, token, body);
  if (dispatched.error) return { ...prepared, execution_status: "backend_error", github_error: dispatched };
  return { ...prepared, execution_status: "completed", approval_required: false, mutation_performed: true, dispatched_workflow: { workflow_id: workflowId, ref: body.ref } };
}

async function prepareReleaseNotes(
  capability: GeneratedCapabilityRuntimeMetadata,
  plan: BackendInvocationPlan,
  params: Record<string, unknown>,
  token: string,
): Promise<Record<string, unknown>> {
  const repo = await repoMetadata(capability, plan, params, token);
  if (repo.error) return repo.error;
  return {
    execution_status: "completed",
    ...metadata(capability, plan),
    result: {
      title: `Release notes for ${repo.owner}/${repo.repo} ${String(params.range ?? "").trim()}`,
      audience: String(params.audience ?? "internal").trim(),
      repository: repoSummary(repo.repoPayload!),
      range: String(params.range ?? "").trim(),
      sections: [
        { title: "Highlights", items: ["Review bounded GitHub context before publishing release notes."] },
        { title: "Governance", items: ["This capability drafts content only and does not create a GitHub release."] },
      ],
    },
    mutation_performed: false,
  };
}

export const backendAdapter = createDefaultBackendAdapter();

File: examples/showcase/github_fronting/custom-code-bundles/github_fronting_go/extensions/backend_adapter.go
package extensions

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"

	"{{ANIP_GO_MODULE_PATH}}/generated"
)

type BackendInvocationContext struct {
	RootPrincipal string
	ApprovalGrant string
}

type BackendAdapter interface {
	Execute(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, adapterInput map[string]any, context BackendInvocationContext) (map[string]any, error)
}

type defaultBackendAdapter struct{}

func CreateDefaultBackendAdapter() BackendAdapter {
	return defaultBackendAdapter{}
}

func (defaultBackendAdapter) Execute(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, context BackendInvocationContext) (map[string]any, error) {
	if len(plan.UnresolvedRequiredBackendInputs) > 0 {
		return map[string]any{
			"execution_status": "backend_input_incomplete",
			"capability_id": capability.CapabilityID,
			"backend_input_contract": plan.BackendInputContract,
			"unresolved_required_backend_inputs": plan.UnresolvedRequiredBackendInputs,
			"note": "Generated host is runnable, but backend-only inputs still require extension completion.",
		}, nil
	}
	token := strings.TrimSpace(os.Getenv("GITHUB_TOKEN"))
	if token == "" {
		return withMetadata(capability, plan, map[string]any{"execution_status": "backend_error", "github_error": map[string]any{"error": "missing_github_token"}}), nil
	}
	switch capability.CapabilityID {
	case "github.repo.search_context":
		return searchRepositoryContext(capability, plan, params, token)
	case "github.issue.prepare":
		return prepareOrCreateIssue(capability, plan, params, token, context)
	case "github.pr.comment.prepare":
		return preparePullRequestComment(capability, plan, params, token, context)
	case "github.workflow.dispatch.request":
		return prepareWorkflowDispatch(capability, plan, params, token, context)
	case "github.release_notes.prepare":
		return prepareReleaseNotes(capability, plan, params, token)
	default:
		return withMetadata(capability, plan, map[string]any{"execution_status": "backend_execution_stub"}), nil
	}
}

func searchRepositoryContext(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, token string) (map[string]any, error) {
	owner, repo := stringParam(params, "owner"), stringParam(params, "repo")
	if !repoAllowed(owner, repo) {
		return restricted(capability, plan, owner, repo), nil
	}
	limit := boundedLimit(params["limit"])
	query := strings.TrimSpace(fmt.Sprintf("repo:%s/%s %s", owner, repo, stringParam(params, "query")))
	payload := githubRequest("GET", "/search/issues?"+url.Values{"q": {query}, "per_page": {strconv.Itoa(limit)}}.Encode(), token, nil)
	if payload["error"] != nil {
		return withMetadata(capability, plan, map[string]any{"execution_status": "backend_error", "github_error": payload}), nil
	}
	items := make([]map[string]any, 0)
	if rawItems, ok := payload["items"].([]any); ok {
		for _, raw := range rawItems {
			item, ok := raw.(map[string]any)
			if !ok || len(items) >= limit {
				continue
			}
			kind := "issue"
			if item["pull_request"] != nil {
				kind = "pull_request"
			}
			items = append(items, map[string]any{
				"number": item["number"],
				"title": item["title"],
				"state": item["state"],
				"html_url": item["html_url"],
				"kind": kind,
			})
		}
	}
	return withMetadata(capability, plan, map[string]any{
		"execution_status": "completed",
		"github_query": query,
		"result": map[string]any{"items": items, "count": len(items), "total_count": payload["total_count"]},
	}), nil
}

func prepareOrCreateIssue(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, token string, context BackendInvocationContext) (map[string]any, error) {
	owner, repo, repoPayload, errResult := repoMetadata(capability, plan, params, token)
	if errResult != nil {
		return errResult, nil
	}
	body := map[string]any{"title": stringParam(params, "title"), "body": stringParam(params, "body")}
	if labels := stringList(params["labels"]); len(labels) > 0 {
		body["labels"] = labels
	}
	if assignees := stringList(params["assignees"]); len(assignees) > 0 {
		body["assignees"] = assignees
	}
	preview := writePreview(capability, plan, "issues.create", fmt.Sprintf("/repos/%s/%s/issues", owner, repo), body, repoPayload)
	if !mutationEnabled(context) {
		return preview, nil
	}
	created := githubRequest("POST", fmt.Sprintf("/repos/%s/%s/issues", url.PathEscape(owner), url.PathEscape(repo)), token, body)
	if created["error"] != nil {
		preview["execution_status"] = "backend_error"
		preview["github_error"] = created
		return preview, nil
	}
	preview["execution_status"] = "completed"
	preview["approval_required"] = false
	preview["mutation_performed"] = true
	preview["created_issue"] = map[string]any{"number": created["number"], "html_url": created["html_url"], "state": created["state"]}
	preview["note"] = "Created GitHub issue after the ANIP runtime validated and reserved an approval grant."
	return preview, nil
}

func preparePullRequestComment(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, token string, context BackendInvocationContext) (map[string]any, error) {
	owner, repo, repoPayload, errResult := repoMetadata(capability, plan, params, token)
	if errResult != nil {
		return errResult, nil
	}
	pullNumber := stringParam(params, "pull_number")
	pull := githubRequest("GET", fmt.Sprintf("/repos/%s/%s/pulls/%s", url.PathEscape(owner), url.PathEscape(repo), url.PathEscape(pullNumber)), token, nil)
	if pull["error"] != nil {
		return withMetadata(capability, plan, map[string]any{"execution_status": "backend_error", "github_error": pull}), nil
	}
	body := map[string]any{"body": strings.TrimSpace(fmt.Sprintf("[%s] %s", firstNonEmptyText(stringParam(params, "comment_purpose"), "triage_update"), stringParam(params, "context")))}
	preview := writePreview(capability, plan, "issues.createComment", fmt.Sprintf("/repos/%s/%s/issues/%s/comments", owner, repo, pullNumber), body, repoPayload)
	preview["pull_request"] = map[string]any{"number": pull["number"], "title": pull["title"], "state": pull["state"]}
	if !mutationEnabled(context) {
		return preview, nil
	}
	posted := githubRequest("POST", fmt.Sprintf("/repos/%s/%s/issues/%s/comments", url.PathEscape(owner), url.PathEscape(repo), url.PathEscape(pullNumber)), token, body)
	if posted["error"] != nil {
		preview["execution_status"] = "backend_error"
		preview["github_error"] = posted
		return preview, nil
	}
	preview["execution_status"] = "completed"
	preview["approval_required"] = false
	preview["mutation_performed"] = true
	preview["posted_comment"] = map[string]any{"id": posted["id"], "html_url": posted["html_url"]}
	return preview, nil
}

func prepareWorkflowDispatch(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, token string, context BackendInvocationContext) (map[string]any, error) {
	owner, repo, repoPayload, errResult := repoMetadata(capability, plan, params, token)
	if errResult != nil {
		return errResult, nil
	}
	workflowID := stringParam(params, "workflow_id")
	body := map[string]any{"ref": stringParam(params, "ref"), "inputs": objectParam(params["inputs"])}
	preview := writePreview(capability, plan, "actions.createWorkflowDispatch", fmt.Sprintf("/repos/%s/%s/actions/workflows/%s/dispatches", owner, repo, workflowID), body, repoPayload)
	if !mutationEnabled(context) {
		return preview, nil
	}
	dispatched := githubRequest("POST", fmt.Sprintf("/repos/%s/%s/actions/workflows/%s/dispatches", url.PathEscape(owner), url.PathEscape(repo), url.PathEscape(workflowID)), token, body)
	if dispatched["error"] != nil {
		preview["execution_status"] = "backend_error"
		preview["github_error"] = dispatched
		return preview, nil
	}
	preview["execution_status"] = "completed"
	preview["approval_required"] = false
	preview["mutation_performed"] = true
	preview["dispatched_workflow"] = map[string]any{"workflow_id": workflowID, "ref": body["ref"]}
	return preview, nil
}

func prepareReleaseNotes(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, token string) (map[string]any, error) {
	owner, repo, repoPayload, errResult := repoMetadata(capability, plan, params, token)
	if errResult != nil {
		return errResult, nil
	}
	releaseRange := stringParam(params, "range")
	return withMetadata(capability, plan, map[string]any{
		"execution_status": "completed",
		"result": map[string]any{
			"title": fmt.Sprintf("Release notes for %s/%s %s", owner, repo, releaseRange),
			"audience": firstNonEmptyText(stringParam(params, "audience"), "internal"),
			"repository": repoSummary(repoPayload),
			"range": releaseRange,
			"sections": []map[string]any{
				{"title": "Highlights", "items": []string{"Review bounded GitHub context before publishing release notes."}},
				{"title": "Governance", "items": []string{"This capability drafts content only and does not create a GitHub release."}},
			},
		},
		"mutation_performed": false,
	}), nil
}

func repoMetadata(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, token string) (string, string, map[string]any, map[string]any) {
	owner, repo := stringParam(params, "owner"), stringParam(params, "repo")
	if !repoAllowed(owner, repo) {
		return owner, repo, nil, restricted(capability, plan, owner, repo)
	}
	payload := githubRequest("GET", fmt.Sprintf("/repos/%s/%s", url.PathEscape(owner), url.PathEscape(repo)), token, nil)
	if payload["error"] != nil {
		return owner, repo, nil, withMetadata(capability, plan, map[string]any{"execution_status": "backend_error", "github_error": payload})
	}
	return owner, repo, payload, nil
}

func githubRequest(method string, path string, token string, body map[string]any) map[string]any {
	var reader io.Reader
	if body != nil {
		content, _ := json.Marshal(body)
		reader = bytes.NewReader(content)
	}
	request, err := http.NewRequest(method, "https://api.github.com"+path, reader)
	if err != nil {
		return map[string]any{"error": "github_request_error", "detail": err.Error()}
	}
	request.Header.Set("Accept", "application/vnd.github+json")
	request.Header.Set("Authorization", "Bearer "+token)
	request.Header.Set("Content-Type", "application/json")
	request.Header.Set("User-Agent", "anip-github-fronting-showcase")
	request.Header.Set("X-GitHub-Api-Version", "2022-11-28")
	response, err := http.DefaultClient.Do(request)
	if err != nil {
		return map[string]any{"error": "github_request_error", "detail": err.Error()}
	}
	defer response.Body.Close()
	raw, _ := io.ReadAll(response.Body)
	payload := map[string]any{}
	if len(raw) > 0 {
		_ = json.Unmarshal(raw, &payload)
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return map[string]any{"error": "github_http_error", "status": response.StatusCode, "detail": payload}
	}
	return payload
}

func withMetadata(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, result map[string]any) map[string]any {
	result["capability_id"] = capability.CapabilityID
	result["selected_backend"] = plan.SelectedBinding
	result["semantic_input"] = plan.SemanticInput
	return result
}

func restricted(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, owner string, repo string) map[string]any {
	return withMetadata(capability, plan, map[string]any{
		"execution_status": "restricted",
		"repository": map[string]any{"owner": owner, "repo": repo},
		"reason": "GitHub repository is outside the configured ANIP repository policy.",
	})
}

func writePreview(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, action string, path string, body map[string]any, repoPayload map[string]any) map[string]any {
	return withMetadata(capability, plan, map[string]any{
		"execution_status": "prepared",
		"approval_required": true,
		"mutation_performed": false,
		"github_action": action,
		"github_metadata": repoSummary(repoPayload),
		"github_request": map[string]any{"method": "POST", "path": path, "body": body},
		"note": "Prepared a GitHub request payload. No GitHub mutation was performed.",
	})
}

func repoSummary(payload map[string]any) map[string]any {
	owner := ""
	if rawOwner, ok := payload["owner"].(map[string]any); ok {
		owner = stringValue(rawOwner["login"])
	}
	return map[string]any{"owner": owner, "repo": payload["name"], "default_branch": payload["default_branch"], "private": payload["private"], "html_url": payload["html_url"]}
}

func repoAllowed(owner string, repo string) bool {
	key := strings.ToLower(owner + "/" + repo)
	blocked := csvSet("ANIP_GITHUB_BLOCKED_REPOS")
	allowed := csvSet("ANIP_GITHUB_ALLOWED_REPOS")
	if blocked[key] {
		return false
	}
	return len(allowed) == 0 || allowed[key]
}

func mutationEnabled(context BackendInvocationContext) bool {
	return os.Getenv("ANIP_GITHUB_ALLOW_MUTATION") == "true" && strings.TrimSpace(context.ApprovalGrant) != ""
}

func csvSet(name string) map[string]bool {
	result := map[string]bool{}
	for _, item := range strings.Split(os.Getenv(name), ",") {
		if value := strings.ToLower(strings.TrimSpace(item)); value != "" {
			result[value] = true
		}
	}
	return result
}

func boundedLimit(value any) int {
	limit, err := strconv.Atoi(stringValue(value))
	if err != nil {
		limit = 20
	}
	if limit < 1 {
		return 1
	}
	if limit > 50 {
		return 50
	}
	return limit
}

func stringList(value any) []string {
	var raw []any
	switch typed := value.(type) {
	case []any:
		raw = typed
	case []string:
		for _, item := range typed {
			raw = append(raw, item)
		}
	case string:
		for _, item := range strings.Split(typed, ",") {
			raw = append(raw, item)
		}
	}
	seen := map[string]bool{}
	result := []string{}
	for _, item := range raw {
		text := strings.TrimSpace(stringValue(item))
		if text != "" && !seen[text] {
			seen[text] = true
			result = append(result, text)
		}
	}
	return result
}

func objectParam(value any) map[string]any {
	if typed, ok := value.(map[string]any); ok {
		return typed
	}
	return map[string]any{}
}

func stringParam(params map[string]any, name string) string {
	return strings.TrimSpace(stringValue(params[name]))
}

func stringValue(value any) string {
	if value == nil {
		return ""
	}
	return fmt.Sprint(value)
}

func firstNonEmptyText(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return strings.TrimSpace(value)
		}
	}
	return ""
}

var BackendAdapterInstance = CreateDefaultBackendAdapter()
